File: public/js/jwysiwyg/help/htmlentity/unicode4jwysiwyg.pl
#!/usr/bin/env perl
use strict;
use warnings;

use Readonly;
use FindBin qw( $Bin );
use File::Spec qw();
use Text::CSV;
use IO::File;
use JSON::XS;

use feature ':5.10';

## These four elements are also valid in XML
## &apos btw is only valid in XML, and is not an HTML entity
## So it works for XHTML documents, but not for HTML/SGML documents
Readonly my @VALID_ESCAPES  => qw[ amp lt gt quot ];
# http://en.wikipedia.org/wiki/List_of_XML_and_HTML_character_entity_references#Character_entity_references_in_HTML
Readonly my $UNICODE_DB     => File::Spec->catfile($Bin, 'db.csv');
# Reserved words of JavaScript should be quoted in Hash
Readonly my @RESERVED_WORDS => qw[ int ];

my $fh = IO::File->new($UNICODE_DB);
my $csv = Text::CSV->new({ binary=>1, sep_char=>"\t", quote_char=>undef, escape_char=>undef });
$csv->column_names( map lc, @{ $csv->getline($fh) } );

## We'll create our own pseudo element for non-matches
my $db = { __replacement => int 65533 };
while ((my $row = $csv->getline_hr($fh))) {
	my ($unicode, $dec) = $row->{'unicode code point (decimal)'} =~ m/ (\S+) \s+ \( (\S+) \) /x;
	$db->{$row->{name}} = int $dec;
}

delete $db->{$_} for @VALID_ESCAPES;

my $json = JSON::XS->new->utf8->encode($db);

## Actually we need not true JSON, but JavaScript valid hash
$json =~ tr/"//d;

for my $word (@RESERVED_WORDS)
{
	my $escaped_word = quotemeta($word);
	$json =~ s/\b$escaped_word\:/"$word":/;
}

print "$json\n";

# Paste in jwysiwyg and you're done.

File: public/js/jwysiwyg/help/htmlentity/unicode4jwysiwyg.rb
#!/usr/bin/env ruby
# encoding: utf-8

# ruby version of Unicode table converter

require 'rubygems'
begin
  require 'bundler/setup'
rescue LoadError
  puts "You should install Bundler using 'gem install bundler' and run 'bundle install'"
end

require 'yajl'
require 'fastercsv'

## These four elements are also valid in XML
## &apos btw is only valid in XML, and is not an HTML entity
## So it works for XHTML documents, but not for HTML/SGML documents
VALID_ESCAPES  = %w[ amp lt gt quot ].freeze
# http://en.wikipedia.org/wiki/List_of_XML_and_HTML_character_entity_references#Character_entity_references_in_HTML
UNICODE_DB     = File.expand_path('./db.csv', File.dirname(__FILE__)).freeze
# Reserved words of JavaScript should be quoted in Hash
RESERVED_WORDS = %w[ int ].freeze

csv = FasterCSV.open(
  UNICODE_DB,
  'rb',
  :col_sep => "\t",
  :row_sep => "\n",
  :quote_char => '`',
  :headers => :first_row
)

## We'll create our own pseudo element for non-matches
db = { '__replacement' => 65533 }

csv.each do |row|
  row['Unicode code point (decimal)'] =~ /(\S+)\s+\((\S+)\)/
  uch, dec, name = $1, $2.to_i, row['Name']
  next if VALID_ESCAPES.include?(name)
  db[name] = dec
end

json = Yajl::Encoder.encode(db)

## Actually we need not true JSON, but JavaScript valid hash
json.gsub!('"', '')

RESERVED_WORDS.each { |word| json.sub!(/\b#{Regexp.escape(word)}\:/, "\"#{word}\":") }

puts json

# Paste in jwysiwyg and you're done.
